src/commonMain/kotlin/fr/acinq/lightning/serialization/Encryption.kt

```diff
@@ -0,0 +1,58 @@
+package fr.acinq.lightning.serialization
+
+import fr.acinq.bitcoin.ByteVector32
+import fr.acinq.bitcoin.Crypto
+import fr.acinq.bitcoin.PrivateKey
+import fr.acinq.lightning.channel.ChannelStateWithCommitments
+import fr.acinq.lightning.channel.PersistedChannelState
+import fr.acinq.lightning.crypto.ChaCha20Poly1305
+import fr.acinq.lightning.utils.runTrying
+import fr.acinq.lightning.utils.toByteVector
+import fr.acinq.lightning.wire.EncryptedChannelData
+
+/**
+ * Utility methods to encrypt/decrypt serialized channel data
+ */
+object Encryption {
+
+    private fun encrypt(key: ByteVector32, bin: ByteArray): ByteArray {
+        // NB: there is a chance of collision here, due to how the nonce is calculated. Probability of collision is once every 2.2E19 times.
+        // See https://en.wikipedia.org/wiki/Birthday_attack
+        val nonce = Crypto.sha256(bin).take(12).toByteArray()
+        val (ciphertext, tag) = ChaCha20Poly1305.encrypt(key.toByteArray(), nonce, bin, ByteArray(0))
+        return ciphertext + nonce + tag
+    }
+
+    private fun decrypt(key: ByteVector32, data: ByteArray): ByteArray {
+        // nonce is 12B, tag is 16B
+        val ciphertext = data.dropLast(12 + 16)
+        val nonce = data.takeLast(12 + 16).take(12)
+        val tag = data.takeLast(16)
+        val plaintext = ChaCha20Poly1305.decrypt(key.toByteArray(), nonce.toByteArray(), ciphertext.toByteArray(), ByteArray(0), tag.toByteArray())
+        return plaintext
+    }
+
+    /**
+     * Convenience method that builds an [EncryptedChannelData] from a [ChannelStateWithCommitments]
+     */
+    fun EncryptedChannelData.Companion.from(key: PrivateKey, state: PersistedChannelState): EncryptedChannelData {
+        val bin = Serialization.serialize(state)
+        val encrypted = encrypt(key.value, bin)
+        // we copy the first 2 bytes as meta-info on the serialization version
+        val data = bin.copyOfRange(0, 2) + encrypted
```

Comment:
I did think about this approach. The hacky part is that I'm copying _two_ bytes, which means that it includes the serialization version (1B), and the discriminator for the channel state (1B). Doing it "cleanly" would have made this creep a bit too deep in the serializers, and hurt the symmetry between serialization and deserialization. Dealing with bytes sort of guarantees that we're not coupling between the encryption layer and the serialization layer.

Reply:
Why do we need to include the discriminator for the channel state? In theory we only need the serialization version, right? We need to make this clearer and cleaner somehow.

> Doing it "cleanly" would have made this creep a bit too deep in the serializers, and hurt the symmetry between serialization and deserialization.

Why? If serialization returns a magic byte and the serialized data, and deserialization takes as input a magic byte (which lets us choose which deserializer version to use) and the serialized data, this is perfectly symmetrical and clean?

Reply:
I gave it a try in 8f35d013, but I think it's not better:
- this meta-info now impacts 3 files instead of one, and adds coupling between those files
- the way I build `SerializedChannelState(bytes[0], bytes[1], bytes)` is still hacky because the data has already been serialized at this point, we cherry pick bytes from the result
- we still need to do the ugly `data.drop(2)` when decrypting
- due to backward compat with previous versions that didn't have the meta-info, we break the symmetry between the `serialize()` and `deserialize()` methods
- I also tried not writing the version byte in `v4/Serialization`, and it started nice, but then it got hacky at the upper layer

Am I missing something?

Reply:
> this meta-info now impacts 3 files instead of one, and adds coupling between those files

I think that coupling between serialization and deserialization is natural, and the coupling with encryption is there anyway since we pad the encrypted result with that meta...

> the way I build SerializedChannelState(bytes[0], bytes[1], bytes) is still hacky because the data has already been serialized at this point, we cherry pick bytes from the result

This is only true for the `stateByte`, not the version one, which is a static value of the serializer. We know that including the state byte is somewhat hackish anyway but may come in handy!

> due to backward compat with previous versions that didn't have the meta-info, we break the symmetry between the serialize() and deserialize() methods

That's because we still do a hacky deserialization method where we try each version in sequence and catch errors. A better deserialization method would start by popping the meta bytes then call the right version decrypter and deserializer depending on the version bytes. But that can be done in a future improvement.

Reply:
As discussed offline, I reverted in 9c528ccd as I think the fix is still unsatisfactory. This doesn't affect serialization and can be revisited in a later PR.